### ISSUE TITLE
Crash in SROA

### ISSUE THREAD

**Issue by arichardson** [CTSRD-CHERI/llvm-project#372]
While reducing std::deque miscompilation I found this SROA assertion error:

```
; RUN: %cheri_purecap_opt -S -inline -sroa -early-cse-memssa %s -o -
source_filename = "/Users/alex/cheri/llvm-project/libcxx/test/std/containers/sequences/deque/deque.cons/iter_iter.pass.cpp"
target datalayout = "E-m:e-pf200:128:128:128:64-i8:8:32-i16:16:32-i64:64-n32:64-S128-A200-P200-G200"
target triple = "mips64c128-unknown-freebsd13-purecap"

%struct.a = type { %struct.b, %struct.b }
%struct.b = type { i32 addrspace(200)* }
declare i32 @baz(...) addrspace(200)
define void @d() addrspace(200) personality i8 addrspace(200)* bitcast (i32 (...) addrspace(200)* @baz to i8 addrspace(200)*) {
  %l = alloca %struct.a, addrspace(200)
  call void @e(%struct.a addrspace(200)* %l)
  unreachable
}
define void @e(%struct.a addrspace(200)* %arg) addrspace(200) {
  %f = alloca %struct.a addrspace(200)*, addrspace(200)
  store %struct.a addrspace(200)* %arg, %struct.a addrspace(200)* addrspace(200)* %f
  %g = load %struct.a addrspace(200)*, %struct.a addrspace(200)* addrspace(200)* %f
  %h = getelementptr %struct.a, %struct.a addrspace(200)* %g, i32 0, i32 0
  %1 = getelementptr %struct.b, %struct.b addrspace(200)* %h
  %i = getelementptr inbounds %struct.a, %struct.a addrspace(200)* %g, i32 0, i32 1
  %j = getelementptr %struct.a, %struct.a addrspace(200)* %g, i32 0, i32 0
  %l = bitcast %struct.b addrspace(200)* %j to i8 addrspace(200)*
  %k = bitcast %struct.b addrspace(200)* %i to i8 addrspace(200)*
  call void @llvm.memcpy.p200i8.p200i8.i64(i8 addrspace(200)* %l, i8 addrspace(200)* %k, i64 32, i1 false)
  ret void
}
; Function Attrs: argmemonly nounwind willreturn
declare void @llvm.memcpy.p200i8.p200i8.i64(i8 addrspace(200)* noalias nocapture writeonly, i8 addrspace(200)* noalias nocapture readonly, i64, i1 immarg) addrspace(200) #0

attributes #0 = { argmemonly nounwind willreturn }
```

```
0.	Program arguments: /Users/alex/cheri/llvm-project/cmake-build-debug/bin/opt -mtriple=cheri-unknown-freebsd -mcpu=cheri128 -mattr=+cheri128 -target-abi purecap -relocation-model pic -S -inline -sroa -early-cse-memssa /Users/alex/cheri/llvm-project/llvm/test/Transforms/SROA/cheri_crash.ll -o - 
1.	Running pass 'CallGraph Pass Manager' on module '/Users/alex/cheri/llvm-project/llvm/test/Transforms/SROA/cheri_crash.ll'.
2.	Running pass 'SROA' on function '@d'
0  libLLVMSupport.dylib     0x000000010ffa460c llvm::sys::PrintStackTrace(llvm::raw_ostream&) + 60
1  libLLVMSupport.dylib     0x000000010ffa4b99 PrintStackTraceSignalHandler(void*) + 25
2  libLLVMSupport.dylib     0x000000010ffa2696 llvm::sys::RunSignalHandlers() + 118
3  libLLVMSupport.dylib     0x000000010ffa77bc SignalHandler(int) + 252
4  libsystem_platform.dylib 0x00007fff748cbb5d _sigtramp + 29
5  libsystem_platform.dylib 0x00000001116e3938 _sigtramp + 18446603343148187128
6  libsystem_c.dylib        0x00007fff747856a6 abort + 127
7  libsystem_c.dylib        0x00007fff7474e20d basename_r + 0
8  libLLVMScalarOpts.dylib  0x000000010e867cd2 llvm::sroa::AllocaSliceRewriter::visitMemTransferInst(llvm::MemTransferInst&) + 1474
9  libLLVMScalarOpts.dylib  0x000000010e866768 llvm::InstVisitor<llvm::sroa::AllocaSliceRewriter, bool>::visitMemCpyInst(llvm::MemCpyInst&) + 40
10 libLLVMScalarOpts.dylib  0x000000010e8662e2 llvm::InstVisitor<llvm::sroa::AllocaSliceRewriter, bool>::delegateCallInst(llvm::CallInst&) + 402
11 libLLVMScalarOpts.dylib  0x000000010e85d23d llvm::InstVisitor<llvm::sroa::AllocaSliceRewriter, bool>::visitCall(llvm::CallInst&) + 29
12 libLLVMScalarOpts.dylib  0x000000010e85c1ff llvm::InstVisitor<llvm::sroa::AllocaSliceRewriter, bool>::visit(llvm::Instruction&) + 1711
13 libLLVMScalarOpts.dylib  0x000000010e85b5ad llvm::InstVisitor<llvm::sroa::AllocaSliceRewriter, bool>::visit(llvm::Instruction*) + 29
14 libLLVMScalarOpts.dylib  0x000000010e848d8a llvm::sroa::AllocaSliceRewriter::visit((anonymous namespace)::Slice const*) + 1194
15 libLLVMScalarOpts.dylib  0x000000010e846eb4 llvm::SROA::rewritePartition(llvm::AllocaInst&, llvm::sroa::AllocaSlices&, llvm::sroa::Partition&) + 2500
16 libLLVMScalarOpts.dylib  0x000000010e849ef4 llvm::SROA::splitAlloca(llvm::AllocaInst&, llvm::sroa::AllocaSlices&) + 1556
17 libLLVMScalarOpts.dylib  0x000000010e84b206 llvm::SROA::runOnAlloca(llvm::AllocaInst&) + 1222
18 libLLVMScalarOpts.dylib  0x000000010e84c7be llvm::SROA::runImpl(llvm::Function&, llvm::DominatorTree&, llvm::AssumptionCache&) + 478
19 libLLVMScalarOpts.dylib  0x000000010e87465d llvm::sroa::SROALegacyPass::runOnFunction(llvm::Function&) + 189
20 libLLVMCore.dylib        0x000000010a93ebe9 llvm::FPPassManager::runOnFunction(llvm::Function&) + 681
21 libLLVMAnalysis.dylib    0x000000010624e5cd (anonymous namespace)::CGPassManager::RunPassOnSCC(llvm::Pass*, llvm::CallGraphSCC&, llvm::CallGraph&, bool&, bool&) + 925
22 libLLVMAnalysis.dylib    0x0000000106248a7d (anonymous namespace)::CGPassManager::RunAllPassesOnSCC(llvm::CallGraphSCC&, llvm::CallGraph&, bool&) + 541
23 libLLVMAnalysis.dylib    0x0000000106247641 (anonymous namespace)::CGPassManager::runOnModule(llvm::Module&) + 433
24 libLLVMCore.dylib        0x000000010a93f928 (anonymous namespace)::MPPassManager::runOnModule(llvm::Module&) + 776
25 libLLVMCore.dylib        0x000000010a93f455 llvm::legacy::PassManagerImpl::run(llvm::Module&) + 293
26 libLLVMCore.dylib        0x000000010a93ff51 llvm::legacy::PassManager::run(llvm::Module&) + 33
27 opt                      0x00000001032dfb83 main + 11267
28 libdyld.dylib            0x00007fff746e03d5 start + 1
/Users/alex/cheri/llvm-project/cmake-build-debug/test/Transforms/SROA/Output/cheri_crash.ll.script: line 1: 76193 Abort trap: 6           /Users/alex/cheri/llvm-project/cmake-build-debug/bin/opt -mtriple=cheri-unknown-freebsd -mcpu=cheri128 -mattr=+cheri128 -target-abi purecap -relocation-model pic -S -inline -sroa -early-cse-memssa /Users/alex/cheri/llvm-project/llvm/test/Transforms/SROA/cheri_crash.ll -o -
```

**Comment by arichardson** [comment]
This was not fixed properly and recent LLVM changes expose the issue again since they infer the alignment for alloca instructions.

**Comment by jrtc27** [comment]
Fixed in c567cfb as noted above.
